### PR TITLE
Leverage Account service for authentication in app service library

### PR DESCRIPTION
**Pull request by Flanker32**
Leverage Account service for authentication in app service library

### CODE REVIEW

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java [hunk 1]
```diff
@@ -37,56 +50,108 @@ public IWebApp webapp(String id) {
     }
 
     public IWebApp webapp(String resourceGroup, String name) {
-        final WebAppEntity webAppEntity = WebAppEntity.builder().resourceGroup(resourceGroup).name(name).build();
+        return webapp(null, resourceGroup, name);
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java:53]: a null subscriptionId? No NPE?

File: azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/SubscriptionScoped.java [hunk 2]
```diff
@@ -44,4 +46,21 @@ public T subscriptions(@Nonnull final List<Subscription> subscriptions) {
...
+
+    public T subscriptions(@Nonnull final String... subscriptions) {
+        assert ArrayUtils.isNotEmpty(subscriptions) : "subscriptions can not be empty!";
+        final List<Subscription> subscriptionList = Azure.az(IAzureAccount.class).account().getSubscriptions().stream()
+                .filter(subscription -> ArrayUtils.contains(subscriptions, subscription.getId()))
+                .collect(Collectors.toList());
+        return this.subscriptions(subscriptionList);
+    }
+
+    // If only one subscription is selected, using it as default subscription
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/SubscriptionScoped.java:62]: we should not have logic of failure to default subscription

File: azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/SubscriptionScoped.java [hunk 3]
```diff
@@ -44,4 +46,21 @@ public T subscriptions(@Nonnull final List<Subscription> subscriptions) {
...
+    public Subscription getDefaultSubscription() {
+        return CollectionUtils.isNotEmpty(subscriptions) && subscriptions.size() == 1 ? subscriptions.get(0) : null;
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/SubscriptionScoped.java:64]: throw exception when subscriptions.size() != 1

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/AppServicePlan.java [hunk 4]
```diff
@@ -24,14 +23,12 @@
 public class AppServicePlan implements IAppServicePlan {
 
     private AppServicePlanEntity entity;
-    private AzureAppService azureAppService;
     private AzureResourceManager azureClient;
     private com.azure.resourcemanager.appservice.models.AppServicePlan appServicePlanInner;
 
-    public AppServicePlan(AppServicePlanEntity appServicePlanEntity, AzureAppService appService) {
+    public AppServicePlan(AppServicePlanEntity appServicePlanEntity, AzureResourceManager azureResourceManager) {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/AppServicePlan.java:29]: rename the second param to azureCLient

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/WebApp.java [hunk 5]
```diff
@@ -41,10 +41,9 @@
     private AzureResourceManager azureClient;
     private com.azure.resourcemanager.appservice.models.WebApp webAppInner;
 
-    public WebApp(WebAppEntity entity, AzureAppService azureAppService) {
+    public WebApp(WebAppEntity entity, AzureResourceManager azureResourceManager) {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/WebApp.java:44]: rename the second param to azureCLient

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java [hunk 6]
```diff
@@ -37,56 +50,108 @@ public IWebApp webapp(String id) {
...
 
-    public AzureResourceManager getAzureResourceManager() {
-        return azureResourceManager;
+    private String getSubscriptionId(@Nullable String resourceId, @Nullable String subscriptionId) {
+        if (StringUtils.isNotEmpty(resourceId)) {
+            return ResourceId.fromString(resourceId).subscriptionId();
+        }
+        if (StringUtils.isEmpty(subscriptionId) && getDefaultSubscription() == null) {
+            throw new AzureToolkitRuntimeException("Subscription id is required for this request.");
+        }
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java:154]: StringUtils.firstNonBlank

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java [hunk 7]
```diff
@@ -37,56 +50,108 @@ public IWebApp webapp(String id) {
...
+        return (httpPipelineCallContext, httpPipelineNextPolicy) -> {
+            final String previousUserAgent = httpPipelineCallContext.getHttpRequest().getHeaders().getValue("User-Agent");
+            httpPipelineCallContext.getHttpRequest().setHeader("User-Agent", String.format("%s %s", userAgent, previousUserAgent));
+            return httpPipelineNextPolicy.process();
+        };
     }
 
-    public AzureResourceManager getAzureResourceManager() {
-        return azureResourceManager;
+    private String getSubscriptionId(@Nullable String resourceId, @Nullable String subscriptionId) {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java:148]: getDefaultSubscriptionId

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/AppServiceUtils.java [hunk 8]
```diff
@@ -167,6 +168,7 @@ static WebAppDeploymentSlotEntity fromWebAppDeploymentSlot(DeploymentSlot deploy
     static AppServicePlanEntity fromAppServicePlan(com.azure.resourcemanager.appservice.models.AppServicePlan appServicePlan) {
         return AppServicePlanEntity.builder()
             .id(appServicePlan.id())
+            .subscriptionId(ResourceId.fromString(appServicePlan.id()).subscriptionId())
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/AppServiceUtils.java:171]: dup code:ResourceId.fromString(appServicePlan.id()).subscriptionId()

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java (hunk 1 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java:53]: getSubscriptionId()

**Review by andxu — CHANGES_REQUESTED** [review]
Please address my comments

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java (hunk 7 above)
Comment (Flanker32) [reply azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java:148]: As we discussed before, entity should include subscription data, no need to refer the default value, so keep its origin name ,just try get subscription from resource id then subscription, and throw exception in other cases

**Review by andxu — DISMISSED** [review]
LGTM

File: azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/SubscriptionScoped.java [hunk 9]
```diff
@@ -44,4 +48,22 @@ public T subscriptions(@Nonnull final List<Subscription> subscriptions) {
...
+        assert ArrayUtils.isNotEmpty(subscriptions) : "subscriptions can not be empty!";
+        final List<Subscription> subscriptionList = Azure.az(IAzureAccount.class).account().getSubscriptions().stream()
+                .filter(subscription -> ArrayUtils.contains(subscriptions, subscription.getId()))
+                .collect(Collectors.toList());
+        return this.subscriptions(subscriptionList);
+    }
+
+    // If only one subscription is selected, using it as default subscription
+    public Subscription getDefaultSubscription() {
+        return Optional.of(CollectionUtils.size(subscriptions) == 1 ? subscriptions.get(0) : null)
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/SubscriptionScoped.java:66]: `subscriptions` is null if user doesn't explicitly scope subscriptions (defaultly scoped to selected subscriptions)

File: azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/SubscriptionScoped.java [hunk 10]
```diff
@@ -44,4 +48,22 @@ public T subscriptions(@Nonnull final List<Subscription> subscriptions) {
...
+                .orElseThrow(() -> new AzureToolkitRuntimeException("Subscription can not be null"));
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/SubscriptionScoped.java:67]: the message is incorrect.

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java [hunk 11]
```diff
@@ -37,56 +50,108 @@ public IWebApp webapp(String id) {
...
+    }
+
+    public IWebApp webapp(String subscriptionId, String resourceGroup, String name) {
+        final WebAppEntity webAppEntity = WebAppEntity.builder().subscriptionId(subscriptionId).resourceGroup(resourceGroup).name(name).build();
         return webapp(webAppEntity);
     }
 
     public IWebApp webapp(WebAppEntity webAppEntity) {
-        return new WebApp(webAppEntity, this);
+        final String subscriptionId = getSubscriptionId(webAppEntity.getId(), webAppEntity.getSubscriptionId());
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java:62]: isn't this interface weird?
pass `subscriptionId` to `getSubscriptionId`?

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/resources/META-INF/services/com.microsoft.azure.toolkit.lib.AzureService [hunk 12]
```diff
@@ -0,0 +1 @@
+com.microsoft.azure.toolkit.lib.appservice.AzureAppService
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/resources/META-INF/services/com.microsoft.azure.toolkit.lib.AzureService:1]: remove?

**Review by wangmingliang-ms — APPROVED** [review]

**Pull request merged**